### DISCUSSION TITLE
Can you have a JS backend?

### DISCUSSION THREAD

**Discussion by trymeouteh** [tauri-apps/tauri#11249] [Q&A] [answered]
Is it possible not use Rust at all and write all the back end code for the app in JavaScript? I know NodeJS, Deno or Bun is not included in Tauri but Tauri does use HTML, CSS and JavaScript and does include JS APIs to access things like the file system.

Is it possible to simply include JS script to the HTML file which will be the back end code for the app?

**Comment by FabianLars** [comment]
> Is it possible to simply include JS script to the HTML file which will be the back end code for the app?

Hmm, somewhat yes? But as you said, there's no nodejs etc so all your JS will be subject to the browser sandbox, so it only works if the sandbox is no problem, or if tauri's extra Apis fulfill your needs.

**Reply by trymeouteh** [reply]
What would be the limitations of using browser sandbox JS with the Tauri extras APIs compared to having NodeJS?

**Reply by FabianLars** [reply, marked as answer]
Well, Tauri's apis don't cover all of the available nodejs apis, and npm packages that rely on nodejs could also not be used. I really can't tell you whether this will be a problem or not for you.

**Comment by sp90** [comment]
I havn't written the code for it or tested if it's possible but it's purely an idea - wouldn't you be able to spawn a "node/bun/deno" instance then make a IPC reverse proxy kinda thing 

So calling to and from your javascript instance would go `browser -> tauri -> js/ts env` and responses `js/ts env -> tauri -> browser`

One thing could be the perf overhead i dont know how much it would cost to "rev proxy those messages"

**Reply by diliprk** [reply]
Lets say I want to a telegram build a login screen with Tauri +NextJS which involves 3 screens.

1. screen 1 - entering telegram mobile nr,  
2. screen 2 -  entering received OTP / passcode, 
3. screen 3 - Enter 2 step verification password (if enabled by user in his telegram account)

I have the backend logic for this built using  `gramJS` package ( `npm install telegram`). I want to know if `Tauri`  can support running these npm packages for hitting telegram APIs, directly on the client side (in the built exe file) ?

**Reply by sp90** [reply]
Well with my approach you don't leave it up to tauri its purely gonna happen in your spawned javascript instance 

Its purely messages back and forth

**Reply by diliprk** [reply]
Can you kindly share your approach with using JS as a backend for Tauri?

Off topic, but do you think that https://deno.com/blog/deno-compile-executable-programs could be an alternative to tauri + Rust?

**Reply by PxlSyl** [reply]
Hi @diliprk , you can use a sidecar or a thread worker, with the rust code calling the node.js part, IDK if it's clear for you but this is working for me (with json args) and IDK your goal and setup so I can't tell if it will feet your needs

**Reply by diliprk** [reply]
Will this sidecar (nodeJS exe backend) load up in the windows taskbar as a new application in another window or will it run silently as a background process?

**Reply by PxlSyl** [reply]
> Will this sidecar (nodeJS exe backend) load up in the windows taskbar as a new application in another window or will it run silently as a background process?

Hi, it's a background process, and if it's some kind of complex node js code, you can set it up to send progress events that will trigger components in the UI/frontend part

**Comment by gianpaj** [comment]
FYI @marcomq created this plugin https://github.com/marcomq/tauri-plugin-deno

And looking at the example, it does what it was asked
https://github.com/marcomq/tauri-plugin-deno/tree/main/examples/tauri-app

**Reply by gianpaj** [reply]
Others are linked from the Plugins' Integrations section https://tauri.app/plugin/#integrations

- https://github.com/kunkunsh/kkrpc - A TypeScript RPC protocol for multiple environments (iframe, web worker, stdio, http, WebSocket)
- https://github.com/typed-sigterm/deno-in-tauri - Run JavaScript/TypeScript code with Deno Code Engine, in Tauri apps.

**Reply by sp90** [reply]
@gianpaj thanks for sharing
